gtsam/linear/tests/testPowerMethod.cpp

```diff
@@ -0,0 +1,105 @@
+
+/**
+ * testPowerMethod.cpp
+ *
+ * @file   testPowerMethod.cpp
+ * @date   Sept 2020
+ * @author Jing Wu
+ * @brief  Check eigenvalue and eigenvector computed by power method
+ */
+
+#include <CppUnitLite/TestHarness.h>
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/VectorSpace.h>
+#include <gtsam/inference/Symbol.h>
+#include <gtsam/linear/GaussianFactorGraph.h>
+#include <gtsam/linear/PowerMethod.h>
+
+#include <Eigen/Core>
+#include <Eigen/Dense>
+#include <Eigen/Eigenvalues>
+#include <iostream>
+#include <random>
+
+using namespace std;
+using namespace gtsam;
+using symbol_shorthand::X;
+
+/* ************************************************************************* */
+TEST(PowerMethod, powerIteration) {
+  // test power iteration, beta is set to 0
+  Sparse A(6, 6);
+  A.coeffRef(0, 0) = 6;
+  A.coeffRef(1, 1) = 5;
+  A.coeffRef(2, 2) = 4;
+  A.coeffRef(3, 3) = 3;
+  A.coeffRef(4, 4) = 2;
+  A.coeffRef(5, 5) = 1;
+  Vector initial = (Vector(6) << 0.24434602, 0.22829942, 0.70094486, 0.15463092, 0.55871359,
+       0.2465342).finished();
+  PowerMethod<Sparse> pf(A, initial);
+  pf.compute(100, 1e-5);
+  EXPECT_LONGS_EQUAL(6, pf.eigenvector().rows());
+
+  Vector6 actual1 = pf.eigenvector();
+  const double ritzValue = actual1.dot(A * actual1);
+  const double ritzResidual = (A * actual1 - ritzValue * actual1).norm();
+  EXPECT_DOUBLES_EQUAL(0, ritzResidual, 1e-5);
+
+  const double ev1 = 6.0;
+  EXPECT_DOUBLES_EQUAL(ev1, pf.eigenvalue(), 1e-5);
+}
+
+/* ************************************************************************* */
+TEST(PowerMethod, useFactorGraph) {
+  // Let's make a scalar synchronization graph with 4 nodes
+  GaussianFactorGraph fg;
+  auto model = noiseModel::Unit::Create(1);
+  for (size_t j = 0; j < 3; j++) {
+    fg.add(X(j), -I_1x1, X(j + 1), I_1x1, Vector1::Zero(), model);
+  }
+  fg.add(X(3), -I_1x1, X(0), I_1x1, Vector1::Zero(), model);  // extra row
+
+  // Get eigenvalues and eigenvectors with Eigen
+  auto L = fg.hessian();
+  Eigen::EigenSolver<Matrix> solver(L.first);
+
+  // Check that we get zero eigenvalue and "constant" eigenvector
+  EXPECT_DOUBLES_EQUAL(0.0, solver.eigenvalues()[0].real(), 1e-9);
+  auto v0 = solver.eigenvectors().col(0);
+  for (size_t j = 0; j < 3; j++) EXPECT_DOUBLES_EQUAL(-0.5, v0[j].real(), 1e-9);
+
+  // find the index of the max eigenvalue
+  size_t maxIdx = 0;
+  for (auto i = 0; i < solver.eigenvalues().rows(); ++i) {
+    if (solver.eigenvalues()(i).real() >= solver.eigenvalues()(maxIdx).real())
+      maxIdx = i;
+  }
+  // Store the max eigenvalue and its according eigenvector
+  const auto ev1 = solver.eigenvalues()(maxIdx).real();
+  auto ev2 = solver.eigenvectors().col(maxIdx).real();
+
+  Vector initial = Vector4::Random();
+  PowerMethod<Matrix> pf(L.first, initial);
+  pf.compute(50, 1e-4);
+  EXPECT_DOUBLES_EQUAL(ev1, pf.eigenvalue(), 1e-8);
+  auto actual2 = pf.eigenvector();
+  EXPECT(assert_equal(ev2, actual2, 3e-5));
```

Comment:
About why using this unittest, the idea comes from that we want to use the graph.hessian() later to extract the A matrix. There are some tests using the EigenSolver from Eigen to see whether the eigenpair result is always "constant" and also to compare the results of power/accelerated method at initial (but now more clear about the correctness of power method). 
But I'm still a little confused here, for the specific matrix A in this unittest, there is no two eigenvectors for one eigenvalue, and the eigenvector is normalized, the only uncertainty here is the sign of the  eigenvector, whether +/-. So I guess it's fine to keep this. I'm not sure. I'm ok with deleting the checking specific eigenvector part.

Reply:
The problem with:

```   EXPECT(assert_equal(ev2, actual2, 3e-5)); ```

is that you are trying to enforce the condition that the (unit) eigenvector ```actual2``` returned by the power method is *exactly* equal to the eigenvector ```ev2``` returned by Eigen.  But there are **always** _at least 2_ unit eigenvectors associated with any given eigenvalue, and in general there will be _infinitely many_ (if that eigenvalue happens to have multiplicity > 1).

So the question here is: why would we _ever_ expect the power method to return _that specific_ eigenvector?

The answer is: we _wouldn't_, because eigenvectors are **never unique**.  And if they're not unique, we shouldn't be trying to enforce uniqueness in a unit test.  This doesn't make mathematical sense.

The correct test of "eigenvector-ness" is the eigenvector equation ```A*x = lambda*x```.  Specifically, we should be checking something like a normalized Ritz residual:

```eps := ||A*x  - lambda*x|| / ||x||```

Note that this error measure does **not** depend upon *which specific* eigenvector we get, since *any* eigenvector ```x``` for eigenvalue ```lambda``` would have ```eps = 0```.

Reply:
Got it!